File: engine/src/warden/registry.py
"""Registry — carrega e indexa configs de projeto de ~/.warden/projects/*.toml.

Separado de config.toml/api_token/warden.db (arquivos únicos, ficam na raiz de
~/.warden) pra não misturar N tomls de projeto com o resto.
"""

from pathlib import Path

from warden.config import ProjectConfig, load_project_config

PROJECTS_DIRNAME = "projects"


class Registry:
    def __init__(self, config_dir: Path):
        self.config_dir = config_dir
        self.projects_dir = config_dir / PROJECTS_DIRNAME
        self._projects: dict[str, ProjectConfig] = {}

    def load(self) -> None:
        self._projects.clear()
        if not self.projects_dir.exists():
            return
        for toml_file in sorted(self.projects_dir.glob("*.toml")):
            project = load_project_config(toml_file)
            self._projects[project.id] = project

    def get(self, project_id: str) -> ProjectConfig:
        return self._projects[project_id]

    def all(self) -> list[ProjectConfig]:
        return list(self._projects.values())

File: agent/Warden.Domain/Registry.cs
using Warden.Domain.Config;

namespace Warden.Domain;

/// <summary>
/// Carrega e indexa configs de projeto de <c>&lt;configDir&gt;/projects/*.toml</c>. Separado de
/// config.toml/api_token/warden.db (arquivos únicos, ficam na raiz de configDir) pra não misturar
/// N tomls de projeto com o resto.
/// </summary>
public sealed class Registry
{
    private const string ProjectsDirName = "projects";

    public string ConfigDir { get; }
    public string ProjectsDir { get; }

    private readonly string _projectsDir;
    private readonly Dictionary<string, ProjectConfig> _projects = new();

    public Registry(string configDir)
    {
        ConfigDir = configDir;
        _projectsDir = Path.Combine(configDir, ProjectsDirName);
        ProjectsDir = _projectsDir;
    }

    public void Load()
    {
        _projects.Clear();
        if (!Directory.Exists(_projectsDir)) return;

        foreach (var tomlFile in Directory.GetFiles(_projectsDir, "*.toml").OrderBy(f => f, StringComparer.Ordinal))
        {
            var project = ConfigLoader.LoadProjectConfig(tomlFile);
            _projects[project.Id] = project;
        }
    }

    public ProjectConfig Get(string projectId) => _projects[projectId];

    public IReadOnlyList<ProjectConfig> All() => _projects.Values.ToList();
}
